File: kits/js/bench/truth-warden.test.js
// Part one of papers/quo-truth.md: what the warden provides. Written from the
// paper alone. Every case here is a sentence of that part made checkable.
import test from 'node:test';
import assert from 'node:assert/strict';
import { Warden, MemoryStore, memoryDelivery, seeds } from '../src/index.js';

const utf8 = new TextEncoder();
const still = () => 1_000;
const random = () => crypto.getRandomValues(new Uint8Array(32));

const COUNTER = `Counter
  bump() int
  read() int
`;

class Counter {
  constructor() {
    this.n = 0n;
  }
  bump() {
    this.n += 1n;
    return this.n;
  }
  read() {
    return this.n;
  }
}

// Two grounds in one process, reached by a delivery that hands bytes straight
// to the far warden's one entry point. No road, no socket, and no step waived.
async function pair() {
  const delivery = memoryDelivery();
  const alice = await Warden.open({ seeds: seeds(random), clock: still, random, delivery });
  const bob = await Warden.open({ seeds: seeds(random), clock: still, random, delivery });
  delivery.attach('mem://alice', alice);
  delivery.attach('mem://bob', bob);
  alice.publish('mem://alice');
  bob.publish('mem://bob');
  return { alice, bob, delivery };
}

test('one entry point takes any arriving bytes and answers bytes or silence', async () => {
  const { alice, bob } = await pair();
  const counter = new Counter();
  const { being } = await alice.hold(counter, { blueprint: COUNTER });
  const invitation = await counter._quo.grant(counter);
  const [handle] = await bob.accept(invitation, { label: 'counter' });

  // An ask arriving is judged and answered.
  assert.equal(await handle.bump(), 1n);
  // Garbage arriving is silence, and the door says nothing about why.
  assert.equal(await alice.arrive(utf8.encode('not an envelope')), null);
  assert.equal(await alice.arrive(new Uint8Array(0)), null);
  assert.ok(being instanceof Uint8Array);
});

test('the closure offers the caller as a fact: holder, rotation or stranger', async () => {
  const { alice, bob } = await pair();
  const seen = [];
  const counter = new Counter();
  counter.bump = function bump() {
    seen.push({ voice: this._quo.caller.voice, kind: this._quo.caller.kind });
    return ++this.n;
  };
  await alice.hold(counter, { blueprint: COUNTER });
  const invitation = await counter._quo.grant(counter);
  const [handle] = await bob.accept(invitation, { label: 'counter' });
  // Accepting is two rotations; the first call after it is a plain ask.
  await handle.bump();
  await handle.bump();
  assert.equal(seen.length, 2);
  assert.equal(seen[0].kind, 'holder');
  assert.ok(seen[0].voice instanceof Uint8Array);
  // A copy, never the row: mutating it changes nothing at the door.
  seen[0].voice.fill(0);
  assert.equal(await handle.bump(), 3n);
});

test('standings are offered as voices only', async () => {
  const { alice, bob } = await pair();
  const counter = new Counter();
  await alice.hold(counter, { blueprint: COUNTER });
  assert.deepEqual(counter._quo.standings(), []);
  const invitation = await counter._quo.grant(counter);
  await bob.accept(invitation, { label: 'counter' });
  const held = counter._quo.standings();
  assert.equal(held.length, 1);
  assert.deepEqual(Object.keys(held[0]), ['voice']);
});

test('grant names the being it opens, and release takes every standing with it', async () => {
  const { alice, bob } = await pair();
  const counter = new Counter();
  const other = new Counter();
  await alice.hold(counter, { blueprint: COUNTER });
  await alice.hold(other, { blueprint: COUNTER });
  const invitation = await counter._quo.grant(other);
  const [handle] = await bob.accept(invitation, { label: 'other' });
  assert.equal(await handle.bump(), 1n);
  // Bob reaches `other` and not `counter`.
  assert.equal(other._quo.standings().length, 1);
  assert.equal(counter._quo.standings().length, 0);
  // Released: Bob's next call meets silence, indistinguishable from anything.
  counter._quo.release(other);
  assert.equal(await handle.bump(), null);
});

test('hold mints a smaller being beside me and relation reaches it through the handle', async () => {
  const { alice } = await pair();
  const counter = new Counter();
  await alice.hold(counter, { blueprint: COUNTER });
  const { handle } = await counter._quo.hold(new Counter(), { blueprint: COUNTER, label: 'small' });
  // Same warden, same shape: asynchronous, a value or silence.
  const answer = handle.bump();
  assert.ok(answer instanceof Promise);
  assert.equal(await answer, 1n);
  assert.equal(await counter._quo.relation('small').read(), 1n);
  counter._quo.release(counter._quo.relation('small'));
  assert.equal(await handle.read(), null);
});

test('why it fell silent is told inward, and nothing crosses the wire', async () => {
  const { alice } = await pair();
  const reasons = [];
  alice.observe((why) => reasons.push(why.reason));
  assert.equal(await alice.arrive(utf8.encode('garbage')), null);
  assert.ok(reasons.length >= 1);
  assert.equal(typeof reasons[0], 'string');
});

test('a hint is stored and carried as an opaque string, never parsed', async () => {
  const { alice, bob } = await pair();
  const counter = new Counter();
  await alice.hold(counter, { blueprint: COUNTER });
  alice.publish('anything at all, even this');
  const invitation = await counter._quo.grant(counter);
  assert.ok(invitation.hints.includes('anything at all, even this'));
  // Delivery walks past what it cannot speak; the door still answers on the
  // road it can.
  const [handle] = await bob.accept(invitation, { label: 'counter' });
  assert.equal(await handle.bump(), 1n);
});

test('what must survive a restart lives in the store the host handed in', async () => {
  const delivery = memoryDelivery();
  const store = new MemoryStore();
  const aliceSeeds = seeds(random);
  let alice = await Warden.open({ seeds: aliceSeeds, clock: still, random, delivery, store });
  const bob = await Warden.open({ seeds: seeds(random), clock: still, random, delivery });
  delivery.attach('mem://alice', alice);
  delivery.attach('mem://bob', bob);
  alice.publish('mem://alice');
  bob.publish('mem://bob');

  const counter = new Counter();
  const beingSeed = random();
  const { being } = await alice.hold(counter, { blueprint: COUNTER, seed: beingSeed });
  const [handle] = await bob.accept(await counter._quo.grant(counter), { label: 'counter' });
  assert.equal(await handle.bump(), 1n);
  const spent = await handle._quo.seal('bump');
  assert.equal(await handle._quo.send(spent), 2n);

  // The process dies. A new warden opens on the same seeds and the same
  // store, holds the same object again, and Bob's standing is still there.
  alice = await Warden.open({ seeds: aliceSeeds, clock: still, random, delivery, store });
  delivery.attach('mem://alice', alice);
  const again = new Counter();
  await alice.hold(again, { blueprint: COUNTER, seed: beingSeed });
  assert.equal(again._quo.standings().length, 1);
  assert.equal(await handle.bump(), 1n);
  // The marks survived too: the envelope spent before the restart is silence.
  assert.equal(await handle._quo.send(spent), null);
  assert.ok(being);
});

test('a label survives a restart onto the row it named, not another at the same door', async () => {
  const delivery = memoryDelivery();
  const store = new MemoryStore();
  const alice = await Warden.open({ seeds: seeds(random), clock: still, random, delivery });
  const bobSeeds = seeds(random);
  let bob = await Warden.open({ seeds: bobSeeds, clock: still, random, delivery, store });
  delivery.attach('mem://alice', alice);
  delivery.attach('mem://bob', bob);
  alice.publish('mem://alice');
  bob.publish('mem://bob');

  const counter = new Counter();
  await alice.hold(counter, { blueprint: COUNTER });
  assert.ok(await bob.knock(alice.card(), { label: 'front' }));
  const [handle] = await bob.accept(await counter._quo.grant(counter), { label: 'counter' });
  assert.equal(await handle.bump(), 1n);
  await bob.keep();

  bob = await Warden.open({ seeds: bobSeeds, clock: still, random, delivery, store });
  delivery.attach('mem://bob', bob);
  const kept = bob.labels.get('counter');
  assert.equal(await kept.handle.bump(), 2n);
  assert.ok(bob.labels.get('front'));
});

// "The graph is nobody's — every being holds only its own edges, so no one
// holds the whole." The whole-graph half of that is a claim about the world and
// no case can hold it: nothing here can speak for every vantage point that
// could ever exist. What is assertable is the mechanism the claim rests on, and
// it is assertable at its strongest point — not what a caller is shown, which
// `subcontractor.test.js` already holds, but what the house itself keeps.
//
// A warden's store snapshot is the fullest vantage that exists anywhere in this
// system: every fact a restart must not lose, including the secrets. So a chain
// of three houses, and the end of it read whole. It is asserted as text rather
// than field by field on purpose — a field added later would slip past a check
// written against the fields of today, and the promise is about the record and
// not about the fields anyone remembered.
test('the fullest vantage in the system stops at one hop: a whole record names no third house', async () => {
  const delivery = memoryDelivery();
  const houses = {};
  for (const name of ['studio', 'agency', 'retailer']) {
    houses[name] = { store: new MemoryStore(), hint: `mem://${name}` };
    houses[name].warden = await Warden.open({
      seeds: seeds(random),
      clock: still,
      random,
      delivery,
      store: houses[name].store,
    });
    delivery.attach(`mem://${name}`, houses[name].warden);
    houses[name].warden.publish(`mem://${name}`);
  }
  const { studio, agency, retailer } = houses;

  // The retailer holds the assets and lets the agency in; the agency holds the
  // brief and lets the studio in. Nobody grants anybody two houses away,
  // because there is no way to.
  const assets = new Counter();
  await retailer.warden.hold(assets, { blueprint: COUNTER, seed: random() });
  await agency.warden.accept(await assets._quo.grant(assets), { label: 'assets' });

  const brief = new Counter();
  await agency.warden.hold(brief, { blueprint: COUNTER, seed: random() });
  await studio.warden.accept(await brief._quo.grant(brief), { label: 'brief' });

  const hex = (bytes) => Buffer.from(bytes).toString('hex');
  const whole = (house) => JSON.stringify(house.store.snapshot);
  const names = (house) => [hex(house.warden.name.pk), hex(house.warden.padlock.pk), house.hint];

  // The middle house holds an edge at each end, and that is the bound rather
  // than a leak. But the two edges are not the same size: the house it reached
  // out to is in its record whole, and the house that reached in is a voice and
  // a way back — its padlock and its hint, never its name. A granting house
  // does not learn who its caller's house is.
  for (const one of names(retailer)) {
    assert.ok(whole(agency).includes(one), 'the agency has lost the house it reached out to');
  }
  assert.ok(whole(agency).includes(hex(studio.warden.padlock.pk)), 'no way back to the studio');
  assert.ok(whole(agency).includes(studio.hint), 'no way back to the studio');
  assert.equal(
    whole(agency).includes(hex(studio.warden.name.pk)),
    false,
    'the agency learned the name of the house that reached in',
  );

  // And the ends know only the middle. Not the far warden's key, not its
  // padlock, not the road it publishes — nothing of it is anywhere in the
  // fullest thing this house holds.
  for (const one of names(retailer)) {
    assert.equal(whole(studio).includes(one), false, `the studio's record names the retailer`);
  }
  for (const one of names(studio)) {
    assert.equal(whole(retailer).includes(one), false, `the retailer's record names the studio`);
  }
});

File: kits/zig/test/truth_warden_test.zig
//! Part one of `papers/quo-truth.md`: what the warden provides. Written from
//! that part alone. Every case here is one of its sentences made checkable.

const std = @import("std");
const warden = @import("warden");
const quo = @import("quo");
const host = @import("host");

const Key = warden.Key;

fn still() i64 {
    return 1_000;
}

var draws: u64 = 0;

fn random() Key {
    // Seeds a bench can repeat. Nothing here is a key: the arithmetic derives
    // every key from what this hands it.
    draws += 1;
    var out: Key = undefined;
    var seed = draws *% 0x9e3779b97f4a7c15;
    for (&out) |*byte| {
        seed = seed *% 6364136223846793005 +% 1442695040888963407;
        byte.* = @truncate(seed >> 33);
    }
    return out;
}

const COUNTER =
    \\Counter
    \\  bump() int
    \\  read() int
    \\
;

const Counter = struct {
    quo: quo.Cell = .{},
    n: i64 = 0,
    seen_voice: ?Key = null,
    seen_kind: ?warden.Kind = null,

    pub fn bump(self: *Counter, at: *quo.At) i64 {
        if (at.caller) |who| {
            self.seen_voice = who.voice;
            self.seen_kind = who.kind;
        }
        self.n += 1;
        return self.n;
    }

    pub fn read(self: *Counter) i64 {
        return self.n;
    }
};

/// Two grounds in one process, each under its own host, reaching the other at
/// distance zero. No socket, and no step waived.
const Pair = struct {
    gpa: std.mem.Allocator,
    threaded: std.Io.Threaded,
    here: host.Host = undefined,
    there: host.Host = undefined,
    alice: *warden.Warden = undefined,
    bob: *warden.Warden = undefined,

    fn init(gpa: std.mem.Allocator, self: *Pair) !void {
        self.* = .{ .gpa = gpa, .threaded = .init(gpa, .{}) };
        const io = self.threaded.io();
        try host.Host.open(gpa, opening(io), &self.here);
        try host.Host.open(gpa, opening(io), &self.there);
        self.alice = &self.here.door;
        self.bob = &self.there.door;
    }

    fn opening(io: std.Io) host.Opening {
        return .{
            .seeds = host.seeds(random),
            .clock = still,
            .random = random,
            .io = io,
            .roads = &.{.memory},
        };
    }

    fn deinit(self: *Pair) void {
        self.here.close();
        self.there.close();
        self.threaded.deinit();
    }
};

fn walk(door: *warden.Warden) quo.At {
    return .{ .door = door, .being = door.name, .gpa = std.testing.allocator };
}

/// Accept an invitation that opens one being, and keep a label beside it. A
/// standing naming exactly one being is what `only` answers.
fn joining(
    door: *warden.Warden,
    gpa: std.mem.Allocator,
    invitation: warden.Invitation,
    label: []const u8,
) !quo.Handle {
    var accepted = (try quo.accepting(door, gpa, invitation, walk(door))).?;
    defer accepted.deinit();
    const handle = accepted.only().?;
    try walk(door).label(label, handle);
    return handle;
}

test "one entry point takes any arriving bytes and answers bytes or silence" {
    const gpa = std.testing.allocator;
    var pair: Pair = undefined;
    try Pair.init(gpa, &pair);
    defer pair.deinit();

    var counter: Counter = .{};
    const held = try quo.holding(pair.alice, Counter, &counter, COUNTER, .{}, gpa);
    const invitation = try held.under.grant(gpa, null);
    defer gpa.free(invitation.hints);

    var handle = try joining(pair.bob, gpa, invitation, "counter");
    try std.testing.expectEqual(@as(?i64, 1), try handle.call(gpa, i64, "bump", .{}));

    // Garbage arriving is silence, and the door says nothing about why.
    try std.testing.expect(pair.alice.arrive(gpa, "not an envelope", null) == null);
    try std.testing.expect(pair.alice.arrive(gpa, "", null) == null);
}

test "the closure offers the caller as a fact: holder, rotation or stranger" {
    const gpa = std.testing.allocator;
    var pair: Pair = undefined;
    try Pair.init(gpa, &pair);
    defer pair.deinit();

    var counter: Counter = .{};
    const held = try quo.holding(pair.alice, Counter, &counter, COUNTER, .{}, gpa);
    const invitation = try held.under.grant(gpa, null);
    defer gpa.free(invitation.hints);
    var handle = try joining(pair.bob, gpa, invitation, "counter");

    // Accepting is two rotations; the first call after it is a plain ask.
    _ = try handle.call(gpa, i64, "bump", .{});
    try std.testing.expectEqual(warden.Kind.holder, counter.seen_kind.?);
    try std.testing.expect(counter.seen_voice != null);
}

test "standings are offered as voices only" {
    const gpa = std.testing.allocator;
    var pair: Pair = undefined;
    try Pair.init(gpa, &pair);
    defer pair.deinit();

    var counter: Counter = .{};
    const held = try quo.holding(pair.alice, Counter, &counter, COUNTER, .{}, gpa);

    const none = try held.under.standings(gpa);
    defer gpa.free(none);
    try std.testing.expectEqual(@as(usize, 0), none.len);

    const invitation = try held.under.grant(gpa, null);
    defer gpa.free(invitation.hints);
    _ = try joining(pair.bob, gpa, invitation, "counter");

    const one = try held.under.standings(gpa);
    defer gpa.free(one);
    try std.testing.expectEqual(@as(usize, 1), one.len);
    // A voice and nothing beside it: `[]Key` carries no mark, no window, no
    // padlock and no hint, and there is nowhere here to put one.
    try std.testing.expect(@TypeOf(one) == []Key);
}

test "grant names the being it opens, and release takes every standing with it" {
    const gpa = std.testing.allocator;
    var pair: Pair = undefined;
    try Pair.init(gpa, &pair);
    defer pair.deinit();

    var counter: Counter = .{};
    var other: Counter = .{};
    const one = try quo.holding(pair.alice, Counter, &counter, COUNTER, .{}, gpa);
    const two = try quo.holding(pair.alice, Counter, &other, COUNTER, .{}, gpa);

    const invitation = try one.under.grant(gpa, two.being());
    defer gpa.free(invitation.hints);
    var handle = try joining(pair.bob, gpa, invitation, "other");
    try std.testing.expectEqual(@as(?i64, 1), try handle.call(gpa, i64, "bump", .{}));

    // Bob reaches `other` and not `counter`.
    const at_two = try two.under.standings(gpa);
    defer gpa.free(at_two);
    const at_one = try one.under.standings(gpa);
    defer gpa.free(at_one);
    try std.testing.expectEqual(@as(usize, 1), at_two.len);
    try std.testing.expectEqual(@as(usize, 0), at_one.len);

    // Released: Bob's next call meets silence, indistinguishable from
    // anything else.
    try std.testing.expect(one.under.release(two.being()));
    try std.testing.expect((try handle.call(gpa, i64, "bump", .{})) == null);
}

test "hold mints a smaller being beside me and relation reaches it through the handle" {
    const gpa = std.testing.allocator;
    var pair: Pair = undefined;
    try Pair.init(gpa, &pair);
    defer pair.deinit();

    var counter: Counter = .{};
    var small: Counter = .{};
    const held = try quo.holding(pair.alice, Counter, &counter, COUNTER, .{}, gpa);
    const beside = try held.under.hold(Counter, &small, COUNTER, "small");

    // Same warden, same shape: leashed, a value or silence, no seal paid.
    try std.testing.expectEqual(@as(?i64, 1), try beside.call(gpa, i64, "bump", .{}));
    const by_label = held.under.relation("small").?;
    try std.testing.expectEqual(@as(?i64, 1), try by_label.call(gpa, i64, "read", .{}));

    try std.testing.expect(held.under.release(beside.being()));
    try std.testing.expect((try beside.call(gpa, i64, "read", .{})) == null);
    try std.testing.expect(held.under.relation("small") == null);
}

const Watcher = struct {
    heard: std.ArrayList([]const u8) = .empty,
    gpa: std.mem.Allocator,

    fn observer(self: *Watcher) warden.Observer {
        return .{ .context = @ptrCast(self), .hush = hush };
    }

    fn hush(context: *anyopaque, reason: []const u8) void {
        const self: *Watcher = @ptrCast(@alignCast(context));
        self.heard.append(self.gpa, reason) catch {};
    }
};

test "why it fell silent is told inward, and nothing crosses the wire" {
    const gpa = std.testing.allocator;
    var pair: Pair = undefined;
    try Pair.init(gpa, &pair);
    defer pair.deinit();

    var watcher: Watcher = .{ .gpa = gpa };
    defer watcher.heard.deinit(gpa);
    pair.alice.observe(watcher.observer());

    try std.testing.expect(pair.alice.arrive(gpa, "garbage", null) == null);
    try std.testing.expect(watcher.heard.items.len >= 1);
}

test "a hint is stored and carried as an opaque string, never parsed" {
    const gpa = std.testing.allocator;
    var pair: Pair = undefined;
    try Pair.init(gpa, &pair);
    defer pair.deinit();

    var counter: Counter = .{};
    const held = try quo.holding(pair.alice, Counter, &counter, COUNTER, .{}, gpa);
    try pair.alice.publishRoad("anything at all, even this");

    const invitation = try held.under.grant(gpa, null);
    defer gpa.free(invitation.hints);
    var carried = false;
    for (invitation.hints) |one| {
        if (std.mem.eql(u8, one, "anything at all, even this")) carried = true;
    }
    try std.testing.expect(carried);

    // Delivery walks past what it cannot speak; the door still answers on the
    // road it can.
    var handle = try joining(pair.bob, gpa, invitation, "counter");
    try std.testing.expectEqual(@as(?i64, 1), try handle.call(gpa, i64, "bump", .{}));
}

test "what delivery is given per row is the way back and nothing else" {
    const gpa = std.testing.allocator;
    var pair: Pair = undefined;
    try Pair.init(gpa, &pair);
    defer pair.deinit();

    var counter: Counter = .{};
    const held = try quo.holding(pair.alice, Counter, &counter, COUNTER, .{}, gpa);
    const invitation = try held.under.grant(gpa, null);
    defer gpa.free(invitation.hints);
    var handle = try joining(pair.bob, gpa, invitation, "counter");
    // The bytes went down through delivery and the answer came back up it, so
    // what follows is about a road that carried rather than an absent one.
    try std.testing.expectEqual(@as(?i64, 1), try handle.call(gpa, i64, "bump", .{}));

    // A padlock and some opaque strings. `warden.Row` has two fields and
    // there is nowhere in it to put a secret.
    try std.testing.expectEqual(@as(usize, 2), @typeInfo(warden.Row).@"struct".fields.len);

    // And the row is all delivery is handed beside the envelope: four
    // parameters, and no key, no voice and no seal among them.
    const Send = @typeInfo(@FieldType(warden.Delivery, "send")).pointer.child;
    const params = @typeInfo(Send).@"fn".params;
    try std.testing.expectEqual(@as(usize, 4), params.len);
    try std.testing.expect(params[2].type.? == warden.Row);
    try std.testing.expect(params[3].type.? == []const u8);
}

test "what must survive a restart lives in the store the host handed in" {
    const gpa = std.testing.allocator;
    var threaded: std.Io.Threaded = .init(gpa, .{});
    defer threaded.deinit();
    const io = threaded.io();

    var kept: host.MemoryStore = .{ .gpa = gpa };
    defer kept.deinit();

    const alice_seeds = host.seeds(random);
    const being_seed = random();

    var there: host.Host = undefined;
    try host.Host.open(gpa, Pair.opening(io), &there);
    defer there.close();
    const bob = &there.door;

    var spent: warden.Warden.Sealed = undefined;
    var handle: quo.Handle = undefined;

    // The ground stands on a road it publishes from its own name, so the one
    // it comes back on after the restart is the one its peer already holds.
    const alice_opening: host.Opening = .{
        .seeds = alice_seeds,
        .clock = still,
        .random = random,
        .io = io,
        .roads = &.{.memory},
        .store = kept.store(),
    };

    {
        var here: host.Host = undefined;
        try host.Host.open(gpa, alice_opening, &here);

        var counter: Counter = .{};
        const held = try quo.holding(
            &here.door,
            Counter,
            &counter,
            COUNTER,
            .{ .seed = being_seed },
            gpa,
        );
        const invitation = try held.under.grant(gpa, null);
        defer gpa.free(invitation.hints);
        handle = try joining(bob, gpa, invitation, "counter");
        try std.testing.expectEqual(@as(?i64, 1), try handle.call(gpa, i64, "bump", .{}));

        spent = (try handle.seal(gpa, "bump", .{})).?;
        try std.testing.expectEqual(@as(?i64, 2), try handle.send(gpa, i64, "bump", spent));
        here.close();
    }

    // The process dies. A new ground opens on the same seeds and the same
    // store, holds the same object again, and Bob's standing is still there.
    var again: host.Host = undefined;
    try host.Host.open(gpa, alice_opening, &again);
    defer again.close();

    var counter: Counter = .{};
    const held = try quo.holding(&again.door, Counter, &counter, COUNTER, .{ .seed = being_seed }, gpa);
    const standing = try held.under.standings(gpa);
    defer gpa.free(standing);
    try std.testing.expectEqual(@as(usize, 1), standing.len);

    // The object is new, so its own count starts again; the standing did not.
    try std.testing.expectEqual(@as(?i64, 1), try handle.call(gpa, i64, "bump", .{}));

    // The marks survived too: the envelope spent before the restart is
    // silence.
    try std.testing.expect((try handle.send(gpa, i64, "bump", spent)) == null);
    spent.deinit(gpa);
}

/// A road that never answers in the response it hands back: whatever it
/// carries returns through the far door on somebody else's thread.
fn carriedOnward(
    _: *anyopaque,
    _: std.mem.Allocator,
    _: warden.Row,
    _: []const u8,
) std.mem.Allocator.Error!warden.Carried {
    return .later;
}

var nowhere: u8 = 0;

fn answersLater() warden.Delivery {
    return .{ .context = @ptrCast(&nowhere), .send = carriedOnward, .later = true };
}

test "a road that answers through the door is refused at open when no platform came with it" {
    const gpa = std.testing.allocator;

    // The refusal is at the opening, not at the ask: a warden with no
    // platform has no thread to bring an answer in and nothing to wait on, so
    // the fault is in what the host handed over and is named there.
    try std.testing.expectError(error.Refused, warden.Warden.open(gpa, .{
        .seeds = host.seeds(random),
        .clock = still,
        .random = random,
        .delivery = answersLater(),
    }));

    // And the same delivery with a platform beside it opens, so what is
    // refused is the pairing and never the road.
    var threaded: std.Io.Threaded = .init(gpa, .{});
    defer threaded.deinit();
    var door = try warden.Warden.open(gpa, .{
        .seeds = host.seeds(random),
        .clock = still,
        .random = random,
        .io = threaded.io(),
        .delivery = answersLater(),
    });
    door.deinit();
}

/// A road that answers in the response it hands back, which is what every
/// road that is one call does.
fn carriedHere(
    _: *anyopaque,
    _: std.mem.Allocator,
    _: warden.Row,
    _: []const u8,
) std.mem.Allocator.Error!warden.Carried {
    return .silence;
}

test "a delivery that answers in its response needs no platform, and opens without one" {
    const gpa = std.testing.allocator;

    var door = try warden.Warden.open(gpa, .{
        .seeds = host.seeds(random),
        .clock = still,
        .random = random,
        .delivery = .{ .context = @ptrCast(&nowhere), .send = carriedHere, .later = false },
    });
    door.deinit();
}
